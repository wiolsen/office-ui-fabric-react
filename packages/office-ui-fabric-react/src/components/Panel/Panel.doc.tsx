import * as React from 'react';
import { IDocPageProps } from '../../common/DocPage.types';
import { PanelCustomExample } from './examples/Panel.Custom.Example';
import { PanelCustomLeftExample } from './examples/Panel.CustomLeft.Example';
import { PanelControlledExample } from './examples/Panel.Controlled.Example';
import { PanelExtraLargeExample } from './examples/Panel.ExtraLarge.Example';
import { PanelFooterExample } from './examples/Panel.Footer.Example';
import { PanelNavigationExample } from './examples/Panel.Navigation.Example';
import { PanelHandleDismissTargetExample } from './examples/Panel.HandleDismissTarget.Example';
import { PanelHiddenOnDismissExample } from './examples/Panel.HiddenOnDismiss.Example';
import { PanelLargeExample } from './examples/Panel.Large.Example';
import { PanelLargeFixedExample } from './examples/Panel.LargeFixed.Example';
import { PanelLightDismissExample } from './examples/Panel.LightDismiss.Example';
import { PanelLightDismissCustomExample } from './examples/Panel.LightDismissCustom.Example';
import { PanelMediumExample } from './examples/Panel.Medium.Example';
import { PanelNonModalExample } from './examples/Panel.NonModal.Example';
import { PanelPreventDefaultExample } from './examples/Panel.PreventDefault.Example';
import { PanelScrollExample } from './examples/Panel.Scroll.Example';
import { PanelSmallFluidExample } from './examples/Panel.SmallFluid.Example';
import { PanelSmallLeftExample } from './examples/Panel.SmallLeft.Example';
import { PanelSmallRightExample } from './examples/Panel.SmallRight.Example';

const PanelSmallRightExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallRight.Example.tsx') as string;
const PanelSmallRightExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallRight.Example.tsx') as string;
const PanelSmallLeftExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallLeft.Example.tsx') as string;
const PanelSmallFluidExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallFluid.Example.tsx') as string;
const PanelMediumExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Medium.Example.tsx') as string;
const PanelLargeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Large.Example.tsx') as string;
const PanelLargeFixedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LargeFixed.Example.tsx') as string;
const PanelExtraLargeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.ExtraLarge.Example.tsx') as string;
const PanelCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Custom.Example.tsx') as string;
const PanelCustomLeftExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.CustomLeft.Example.tsx') as string;
const PanelControlledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Controlled.Example.tsx') as string;
const PanelHiddenOnDismissExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.HiddenOnDismiss.Example.tsx') as string;
const PanelLightDismissExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismiss.Example.tsx') as string;
const PanelLightDismissCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismissCustom.Example.tsx') as string;
const PanelLightDismissCustomExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismissCustom.Example.tsx') as string;
const PanelNonModalExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.NonModal.Example.tsx') as string;
const PanelFooterExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Footer.Example.tsx') as string;
const PanelNavigationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Navigation.Example.tsx') as string;
const PanelPreventDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.PreventDefault.Example.tsx') as string;
const PanelHandleDismissTargetExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.HandleDismissTarget.Example.tsx') as string;
const PanelScrollExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Scroll.Example.tsx') as string;

export const PanelPageProps: IDocPageProps = {
  title: 'Panel',
  componentName: 'Panel',
  componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Panel',
  examples: [
    {
      title: 'Panel - Small Panel, Anchored Right, Fixed Width',
      code: PanelSmallRightExampleCode,
      view: <PanelSmallRightExample />,
      codepenJS: PanelSmallRightExampleCodepen
    },
    {
      title: 'Panel - Small Panel, Anchored Left, Fixed Width',
      code: PanelSmallLeftExampleCode,
      view: <PanelSmallLeftExample />
    },
    {
      title: 'Panel - Small Panel, Full Screen, Fluid Width',
      code: PanelSmallFluidExampleCode,
      view: <PanelSmallFluidExample />
    },
    {
      title: 'Panel - Controlled Visibility',
      code: PanelControlledExampleCode,
      view: <PanelControlledExample />
    },
    {
      title: 'Panel - Medium',
      code: PanelMediumExampleCode,
      view: <PanelMediumExample />
    },
    {
      title: 'Panel - Large',
      code: PanelLargeExampleCode,
      view: <PanelLargeExample />
    },
    {
      title: 'Panel - LargeFixed',
      code: PanelLargeFixedExampleCode,
      view: <PanelLargeFixedExample />
    },
    {
      title: 'Panel - Extra Large',
      code: PanelExtraLargeExampleCode,
      view: <PanelExtraLargeExample />
    },
    {
      title: 'Panel - Custom',
      code: PanelCustomExampleCode,
      view: <PanelCustomExample />
    },
    {
      title: 'Panel - Custom Width, Anchored Left',
      code: PanelCustomLeftExampleCode,
      view: <PanelCustomLeftExample />
    },
    {
      title: 'Panel - Hidden on Dismiss',
      code: PanelHiddenOnDismissExampleCode,
      view: <PanelHiddenOnDismissExample />
    },
    {
      title: 'Panel - Light Dismiss',
      code: PanelLightDismissExampleCode,
      view: <PanelLightDismissExample />
    },
    {
      title: 'Panel - Custom Light Dismiss',
      code: PanelLightDismissCustomExampleCode,
      codepenJS: PanelLightDismissCustomExampleCodepen,
      view: <PanelLightDismissCustomExample />
    },
    {
      title: 'Panel - Non-Modal',
      code: PanelNonModalExampleCode,
      view: <PanelNonModalExample />
    },
    {
      title: 'Panel - Footer',
      code: PanelFooterExampleCode,
      view: <PanelFooterExample />
    },
    {
      title: 'Panel - Navigation',
      code: PanelNavigationExampleCode,
      view: <PanelNavigationExample />
    },
    {
      title: 'Panel - Prevent Default Sample',
      code: PanelPreventDefaultExampleCode,
      view: <PanelPreventDefaultExample />
    },
    {
      title: 'Panel - Handle Dismiss Target Sample',
      code: PanelHandleDismissTargetExampleCode,
      view: <PanelHandleDismissTargetExample />
    },
    {
      title: 'Panel - Scrolling Content Sample',
      code: PanelScrollExampleCode,
      view: <PanelScrollExample />
    },
    {
      title: 'Panel - Scrolling Content Sample',
      code: PanelScrollExampleCode,
      view: <PanelScrollExample />
    }
  ],
  overview: require<string>('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelOverview.md'),
  bestPractices: '',
  dos: require<string>('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelDos.md'),
  donts: require<string>('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelDonts.md'),
  isHeaderVisible: true,
  isFeedbackVisible: true,
  allowNativeProps: true
};
