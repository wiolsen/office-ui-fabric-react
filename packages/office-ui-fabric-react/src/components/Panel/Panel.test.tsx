import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { PanelBase } from './Panel.base';
import { Panel } from './Panel';
let div: HTMLElement;

const ReactDOM = require('react-dom');

describe('Panel', () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders Panel correctly', () => {
    // Mock createPortal to capture its component hierarchy in snapshot output.
    ReactDOM.createPortal = jest.fn(element => {
      return element;
    });

    const component = renderer.create(
      <Panel isOpen={true} headerText="Test Panel">
        <span>Content goes here</span>
      </Panel>
    );

    expect(component.toJSON()).toMatchSnapshot();

    ReactDOM.createPortal.mockClear();
  });

  describe('open', () => {
    beforeEach(() => {
      div = document.createElement('div');
    });

    afterEach(() => {
      ReactDOM.unmountComponentAtNode(div);
    });

    it('fires the correct events', () => {
      let openedCalled = false;
      let openCalled = false;
      let dismissedCalled = false;
      let dismissCalled = false;
      let dismissCount = 0;

      const setOpenTrue = (): void => {
        openCalled = true;
      };
      const setOpenedTrue = (): void => {
        openedCalled = true;
      };
      const setDismissTrue = (): void => {
        dismissCalled = true;
        dismissCount++;
      };
      const setDismissedTrue = (): void => {
        dismissedCalled = true;
      };
      jest.useFakeTimers();

      const panel: PanelBase = ReactDOM.render(
        <PanelBase onOpen={setOpenTrue} onOpened={setOpenedTrue} onDismiss={setDismissTrue} onDismissed={setDismissedTrue} />,
        div
      ) as any;

      panel.open();

      expect(openCalled).toEqual(true);
      expect(openedCalled).toEqual(false);

      jest.runOnlyPendingTimers();

      expect(openedCalled).toEqual(true);

      expect(dismissCalled).toEqual(false);
      expect(dismissedCalled).toEqual(false);

      panel.dismiss();

      expect(dismissCalled).toEqual(true);
      expect(dismissedCalled).toEqual(false);

      // Dismiss should only be called once per dismiss.
      expect(dismissCount).toEqual(1);

      jest.runOnlyPendingTimers();

      expect(dismissedCalled).toEqual(true);
    });
  });

  describe('onClose', () => {
    beforeEach(() => {
      div = document.createElement('div');
    });

    afterEach(() => {
      ReactDOM.unmountComponentAtNode(div);
    });
  });
});
