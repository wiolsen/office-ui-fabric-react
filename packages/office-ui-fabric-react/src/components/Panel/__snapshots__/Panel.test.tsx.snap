// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel renders Panel correctly 1`] = `
<span
  className="ms-layer"
>
  <div
    className=
        ms-Fabric
        ms-Layer-content
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #333333;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          visibility: visible;
        }
        & button {
          font-family: inherit;
        }
        & input {
          font-family: inherit;
        }
        & textarea {
          font-family: inherit;
        }
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDoubleClick={[Function]}
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragEnter={[Function]}
    onDragExit={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onSubmit={[Function]}
  >
    <div
      aria-labelledby="Panel0-headerText"
      aria-modal="true"
      className=


      onKeyDown={[Function]}
      role="dialog"
      style={
        Object {
          "outline": "none",
          "overflowY": undefined,
        }
      }
    >
      <div
        aria-hidden={false}
        className=
            ms-Panel
            is-open
            ms-Panel--hasCloseButton
            {
              -moz-osx-font-smoothing: grayscale;
              -webkit-font-smoothing: antialiased;
              bottom: 0px;
              font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
              font-size: 14px;
              font-weight: 400;
              left: 0px;
              pointer-events: none;
              position: absolute;
              right: 0px;
              top: 0px;
            }
      >
        <div
          className=
              ms-Overlay
              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                background-color: rgba(255,255,255,.4);
                bottom: 0px;
                cursor: pointer;
                font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                font-size: 14px;
                font-weight: 400;
                left: 0px;
                pointer-events: auto;
                position: absolute;
                right: 0px;
                top: 0px;
              }
              @media screen and (-ms-high-contrast: active){& {
                border: 1px solid WindowText;
                opacity: 0;
              }
        />
        <div
          className=
              ms-Panel-main
              {
                -webkit-overflow-scrolling: touch;
                background-color: #ffffff;
                bottom: 0px;
                box-shadow: 0px 0px 30px 0px rgba(0,0,0,0.2);
                display: flex;
                flex-direction: column;
                left: auto;
                max-height: 768px;
                overflow-x: hidden;
                overflow-y: auto;
                pointer-events: auto;
                position: absolute;
                right: 0px;
                top: 0px;
                width: 100%;
              }
              @media screen and (-ms-high-contrast: active){& {
                border-left: 3px solid #eaeaea;
                border-right: 3px solid #eaeaea;
              }
              @media (min-width: 480px){& {
                width: 340px;
              }
          onBlur={[Function]}
          onFocus={[Function]}
          onFocusCapture={[Function]}
          style={Object {}}
        >
          <div
            data-is-visible={true}
            onFocus={[Function]}
            style={
              Object {
                "pointerEvents": "none",
                "position": "fixed",
              }
            }
            tabIndex={0}
          />
          <div
            className=
                ms-Panel-commands

            data-is-visible={true}
          >
            <div
              className=
                  ms-Panel-navigation
                  {
                    display: flex;
                    height: 44px;
                    justify-content: flex-end;
                    padding-bottom: 0;
                    padding-left: 5px;
                    padding-right: 5px;
                    padding-top: 0;
                  }
            >
              <button
                className=
                    ms-Button
                    ms-Button--icon
                    ms-Panel-closeButton
                    ms-PanelAction-close
                    {
                      -moz-osx-font-smoothing: grayscale;
                      -webkit-font-smoothing: antialiased;
                      background-color: transparent;
                      border-radius: 0px;
                      border: none;
                      box-sizing: border-box;
                      color: #666666;
                      cursor: pointer;
                      display: inline-block;
                      font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                      font-size: 20px;
                      font-weight: 400;
                      height: auto;
                      margin-bottom: 0px;
                      margin-left: 0px;
                      margin-right: 0px;
                      margin-top: 0px;
                      outline: transparent;
                      padding-bottom: 0;
                      padding-left: 4px;
                      padding-right: 4px;
                      padding-top: 0;
                      position: relative;
                      text-align: center;
                      text-decoration: none;
                      user-select: none;
                      vertical-align: top;
                      width: 44px;
                    }
                    &::-moz-focus-inner {
                      border: 0;
                    }
                    .ms-Fabric--isFocusVisible &:focus:after {
                      border: 1px solid #ffffff;
                      bottom: 0px;
                      content: "";
                      left: 0px;
                      outline: 1px solid #666666;
                      position: absolute;
                      right: 0px;
                      top: 0px;
                      z-index: 1;
                    }
                    @media screen and (-ms-high-contrast: active){.ms-Fabric--isFocusVisible &:focus:after {
                      border: none;
                      bottom: -2px;
                      left: -2px;
                      outline-color: ButtonText;
                      right: -2px;
                      top: -2px;
                    }
                    &:active > * {
                      left: 0px;
                      position: relative;
                      top: 0px;
                    }
                    &:hover {
                      color: #333333;
                    }
                    @media screen and (-ms-high-contrast: active){&:hover {
                      border-color: Highlight;
                      color: Highlight;
                    }
                    &:active {
                      color: #0078d4;
                    }
                data-is-focusable={true}
                data-is-visible={true}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                type="button"
              >
                <div
                  className=
                      ms-Button-flexContainer
                      {
                        align-items: center;
                        display: flex;
                        flex-wrap: nowrap;
                        height: 100%;
                        justify-content: center;
                      }
                >
                  <i
                    aria-hidden={true}
                    className=
                        ms-Button-icon
                        {
                          -moz-osx-font-smoothing: grayscale;
                          -webkit-font-smoothing: antialiased;
                          display: inline-block;
                          flex-shrink: 0;
                          font-family: "FabricMDL2Icons";
                          font-size: 16px;
                          font-style: normal;
                          font-weight: normal;
                          height: 16px;
                          line-height: 16px;
                          margin-bottom: 0;
                          margin-left: 4px;
                          margin-right: 4px;
                          margin-top: 0;
                          speak: none;
                          text-align: center;
                          vertical-align: middle;
                        }
                    data-icon-name="Cancel"
                    role="presentation"
                  >
                    îœ‘
                  </i>
                </div>
              </button>
            </div>
          </div>
          <div
            className=
                ms-Panel-contentInner
                {
                  display: flex;
                  flex-direction: column;
                  flex-grow: 1;
                  max-height: 768px;
                  overflow-y: hidden;
                }
          >
            <div
              className=
                  ms-Panel-header
                  {
                    flex-grow: 0;
                    margin-bottom: 14px;
                    margin-left: 0;
                    margin-right: 0;
                    margin-top: 14px;
                    padding-left: 16px;
                    padding-right: 16px;
                  }
                  @media screen and (min-width: 640px){& {
                    padding-left: 32px;
                    padding-right: 32px;
                  }
                  @media screen and (min-width: 1366px){& {
                    padding-left: 40px;
                    padding-right: 40px;
                  }
                  @media (min-width: 1024px){& {
                    margin-top: 30px;
                  }
            >
              <p
                aria-level={2}
                className=
                    ms-Panel-headerText
                    {
                      -moz-osx-font-smoothing: grayscale;
                      -webkit-font-smoothing: antialiased;
                      color: #333333;
                      font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                      font-size: 21px;
                      font-weight: 100;
                      line-height: 32px;
                      margin-bottom: 0px;
                      margin-left: 0px;
                      margin-right: 0px;
                      margin-top: 0px;
                    }
                id="Panel0-headerText"
                role="heading"
              >
                Test Panel
              </p>
            </div>
            <div
              className=
                  ms-Panel-scrollableContent
                  {
                    height: 768px;
                    overflow-y: auto;
                  }
              data-is-scrollable={true}
            >
              <div
                className=
                    ms-Panel-content
                    {
                      margin-bottom: 0px;
                      padding-bottom: 20px;
                      padding-left: 16px;
                      padding-right: 16px;
                    }
                    @media screen and (min-width: 640px){& {
                      padding-left: 32px;
                      padding-right: 32px;
                    }
                    @media screen and (min-width: 1366px){& {
                      padding-left: 40px;
                      padding-right: 40px;
                    }
              >
                <span>
                  Content goes here
                </span>
              </div>
            </div>
          </div>
          <div
            data-is-visible={true}
            onFocus={[Function]}
            style={
              Object {
                "pointerEvents": "none",
                "position": "fixed",
              }
            }
            tabIndex={0}
          />
        </div>
      </div>
    </div>
  </div>
</span>
`;
